lib/rspec/mocks/proxy.rb

```diff
@@ -166,22 +174,24 @@ def message_received(message, *args, &block)
           end
           stub.invoke(nil, *args, &block)
         elsif expectation
+          expectation.unadvise(messages_arg_list) if defined? expectation.similar_messages
```

Comment:
The `defined?` check is a code smell (especially since the relationship between `unadvise` and `similar_messages` is not obvious).  Instead, I think you can define `unadvise` as a no-op on `RSpec::Mocks::SimpleMessageExpectation`.